### ISSUE TITLE
ParityDb migration tests are flaky

### ISSUE THREAD

**Issue by altonen** [paritytech/polkadot#7609]
PR https://github.com/paritytech/polkadot/pull/7601 has now failed twice on `chains_db::upgrade::tests::test_paritydb_migrate_2_to_3` test. The issue is also reproducible locally. If the tests are run in a loop, both `test_paritydb_migrate_2_to_3` and `test_paritydb_migrate_1_to_2` fail but not consistently.

The error that is printed:
```
---- parachains_db::upgrade::tests::test_paritydb_migrate_1_to_2 stdout ----
thread 'parachains_db::upgrade::tests::test_paritydb_migrate_1_to_2' panicked at 'called `Result::unwrap()` on an `Err` value: Locked(Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" })', node/service/src/parachains_db/upgrade.rs:387:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

These tests failures are also observable on master. Bisect revealed https://github.com/paritytech/polkadot/pull/7337 to be the PR that introduced regression. cc @mrcnski @s0me0ne-unkn0wn

**Comment by ggwpez** [comment]
Yea we may need to comment those if they impede the development workflow.

**Comment by mrcnski** [comment]
Trying to reproduce but unsuccessful so far. What command did you run? I'm doing

```sh
while cargo ltest test_paritydb_migrate_2_to_3; do :; done
```

**Comment by mrcnski** [comment]
OK I get it if I just do `while cargo ltest; [...]`. Weird.

**Comment by altonen** [comment]
> Trying to reproduce but unsuccessful so far. What command did you run? I'm doing
> 
> ```shell
> while cargo ltest test_paritydb_migrate_2_to_3; do :; done
> ```

running
```bash
while true; do cargo test || break; done
```

in `node/service/src/parachain_db` worked for me

**Comment by mrcnski** [comment]
I'm flummoxed, I don't see how these tests can be related to #7337. The tests and the error seem quite specific to paritydb.

```
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: Other, error: "Error removing COL_SESSION_WINDOW_DATA Locked(Os { code: 35, kind: WouldBlock, message: \"Resource temporarily unavailable\" })" })
```

**Comment by mrcnski** [comment]
I've confirmed locally the issue was introduced by #7337. Weird.

**Comment by ordian** [comment]
> If I understand correctly this only happens when there are multuple tests running? The error message here means that the database is already open. This could happen if mutiple tests use the same temp dir.
> 
> In any case, these upgrade tests are probably obsolete. I'd check with the polkadot team if it even makes sense to still maintain the upgrade code. cc @ordian @sandreim

Indeed, the test can be removed now.

**Comment by mrcnski** [comment]
Yeah I'll go bisect the branch. Partly out of morbid curiosity and partly because I'm afraid there is a deeper issue that could reappear later.

**Comment by mrcnski** [comment]
The test failures were introduced in dc5d6a0ed0 where we added some other tests in the same crate. The new tests could not run at the same time as each other[^1], so I had added some code like this:

[^1]: though in retrospect there is a slight tweak that could be made so they don't step on each other's toes.

```rust
// Ensure tests run one at a time so they don't interfere with each other.
static OVERRIDES_LOCK: Mutex<()> = Mutex::new(());
let _lock = OVERRIDES_LOCK.lock()?;
```

in the next commit this was replaced with the [`serial_test`](https://docs.rs/serial_test/latest/serial_test/) crate which does the same thing and the test failures persisted.

My guess is that this affected the order that tests are run in, in such a way that caused paritydb tests to step on each other's toes. 

Question: is this is expected behavior in paritydb or a bug that could manifest in production? I guess in production there is only ever one instance of paritydb? Maybe we should at least catch this error and throw a more descriptive one for the benefit of future devs? @ordian

**Comment by arkpar** [comment]
> Question: is this is expected behavior in paritydb or a bug that could manifest in production? I guess in production there is only ever one instance of paritydb? Maybe we should at least catch this error and throw a more descriptive one for the benefit of future devs? @ordian

Mutiple database instances should not interfere unless they use the same path. Each test must use a unique temp dir.

**Comment by arkpar** [comment]
I don't see anything wrong with the tests. They should have worked. I've looked at the `tempfile` crate to see if it could produce the same path if called simulatneously from two different threads, but it seems to be allright.

I've tried reproducing the issue with `while true; do cargo test || break; done`, on macos but was unable to. So maybe the issue is platform dependent. I'd add some logs to see if the paths are indeeed different. setting `RUST_LOG=parity-db=debug` and running `cargo test --release -- --nocapture` could also shed some light.
